service/src/java/org/apache/hive/service/cli/operation/hplsql/HplSqlQueryExecutor.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.hive.hplsql.executor.QueryResult;
+import org.apache.hive.hplsql.executor.RowResult;
+import org.apache.hive.service.cli.ColumnDescriptor;
+import org.apache.hive.service.cli.FetchOrientation;
+import org.apache.hive.service.cli.FetchType;
+import org.apache.hive.service.cli.HiveSQLException;
+import org.apache.hive.service.cli.OperationHandle;
+import org.apache.hive.service.cli.RowSet;
+import org.apache.hive.service.cli.TableSchema;
+import org.apache.hive.service.cli.session.HiveSession;
+
+/**
+ * Executing HiveQL from HPL/SQL directly, without JDBC or Thrift.
+ */
+public class HplSqlQueryExecutor implements QueryExecutor {
+  public static final String QUERY_EXECUTOR = "QUERY_EXECUTOR";
+  public static final String HPLSQL = "HPLSQL";
+  private final HiveSession hiveSession;
+  private long fetchSize;
+
+  public HplSqlQueryExecutor(HiveSession hiveSession) {
+    this.fetchSize = hiveSession.getHiveConf().getIntVar(HiveConf.ConfVars.HIVE_SERVER2_THRIFT_RESULTSET_DEFAULT_FETCH_SIZE);
+    this.hiveSession = hiveSession;
+  }
+
+  @Override
+  public QueryResult executeQuery(String sql, ParserRuleContext ctx) {
+    try {
+      Map<String, String> confOverlay = new HashMap<>();
+      confOverlay.put(QUERY_EXECUTOR, HPLSQL);
+      OperationHandle operationHandle = hiveSession.executeStatement(sql, confOverlay);
+      return new QueryResult(new OperationRowResult(operationHandle), () -> metadata(operationHandle), null);
+    } catch (HiveSQLException e) {
+      return new QueryResult(null, () -> new Metadata(Collections.emptyList()), e);
+    }
+  }
+
+  public Metadata metadata(OperationHandle operationHandle) {
+    try {
+      TableSchema meta = hiveSession.getResultSetMetadata(operationHandle);
+      List<ColumnMeta> colMeta = new ArrayList<>();
+      for (int i = 0; i < meta.getSize(); i++) {
+        ColumnDescriptor col = meta.getColumnDescriptorAt(i);
+        colMeta.add(new ColumnMeta(col.getName(), col.getTypeName(), col.getType().toJavaSQLType()));
+      }
+      return new Metadata(colMeta);
+    } catch (HiveSQLException e) {
+      throw new QueryException(e);
+    }
+  }
+
+  private class OperationRowResult implements RowResult {
+    private final OperationHandle handle;
+    private RowSet rows;
+    private Iterator<Object[]> iterator;
+    private Object[] current;
+
+    private OperationRowResult(OperationHandle operationHandle) {
+      this.handle = operationHandle;
+    }
+
+    @Override
+    public boolean next() {
+      if (rows == null) {
+        this.rows = fetch();
+        this.iterator = rows.iterator();
+      }
+      if (iterator.hasNext()) {
+        current = iterator.next();
+        return true;
+      } else {
+        current = null;
+        return false;
+      }
+    }
+
+    private RowSet fetch() {
+      try {
+        return hiveSession.fetchResults(
+                handle, FetchOrientation.FETCH_NEXT, fetchSize, FetchType.QUERY_OUTPUT);
```

Comment:
I think if I set fetchsize to 1 - and then I try to fetch a resultset of 10 rows with hplsql; I'll get 1 row back

Reply:
Right. I think I fixed it by adding an extra condition.